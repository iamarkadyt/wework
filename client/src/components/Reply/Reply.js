import React, { Component } from 'react'
import { connect } from 'react-redux'
import './Reply.scss'

import Field from '../Field/Field'
import { fetchUsersStats } from '../../state/actions/userActions'

class Reply extends Component {
    constructor(props) {
        super(props)
        this.state = {}
    }

    handleSubmit() {
        this.setState({ text: '' })
        this.props.fetchUsersStats()
    }

    render() {
        const { onSubmit, flat, rows, errors, authedUser } = this.props
        const flatStyle = {
            borderRadius: 'unset',
            boxShadow: 'unset',
            marginTop: 'unset',
        }

        return <form
            className="Reply-container"
            style={flat ? flatStyle : null}
            onSubmit={e => {
                e.preventDefault()
                onSubmit(this.state, this.handleSubmit.bind(this))
            }}>
            <p className="Reply-header">Post as <b>{authedUser.name}:</b></p>
            <Field
                type="textarea"
                rows={rows || "3"}
                style={{ fontSize: '1.3rem' }}
                value={this.state.text}
                onChange={e => this.setState({ text: e.target.value })}
                error={errors.text} />
            <Field
                type="submit"
                containerStyle={{ marginTop: '.5rem' }}
                label="Submit" />
        </form>
    }
}

export const mapStateToProps = state => ({
    authedUser: state.user,
    errors: state.err.formErrors
})

export const _UnconnectedReply = Reply
export default connect(mapStateToProps, { fetchUsersStats })(Reply)
