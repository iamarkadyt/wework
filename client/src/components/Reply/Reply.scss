@import '../../variables.scss';
.Reply-container {
    width: 100%;
    margin-top: 1rem;
    display: flex;
    flex-direction: column;
    align-items: center;
    box-shadow: $box_shadow_main;
    background-color: $color-bg-secondary;
    color: $color-font_main;
    border-radius: .1rem;
    padding: 0 1rem 1rem 1rem;
    textarea {
        font-size: 1.2rem;
        width: 100%;
    }
    button[type=submit] {
        font-size: 1.2rem;
        align-self: flex-end;
    }
}

.Reply-header {
    align-self: start;
    margin-top: 1rem;
    margin-left: .2rem;
    margin-bottom: -.4rem;
    font-size: 1.2rem;
}
