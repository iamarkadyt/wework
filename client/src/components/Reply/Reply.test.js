import React from 'react'
import { shallow } from 'enzyme'
import { _UnconnectedReply as Reply } from './Reply'
import cloneDeep from 'lodash.clonedeep'

let mountedComponent, props

const getMockProps = () => {
  return cloneDeep({
    onSubmit: jest.fn(),
    flat: true,
    rows: 5,
    errors: {},
    authedUser: { name: 'Name' },
    fetchUsersStats: jest.fn()
  })
}

const comp = () => {
  if (!mountedComponent) {
    mountedComponent = shallow(<Reply {...props} />)
  }
  return mountedComponent
}

describe('Reply', () => {
  beforeEach(() => {
    mountedComponent = undefined
    props = getMockProps()
  })

  describe('props', () => {
    describe('effects', () => {
      describe('flat', () => {
        it('if true, flatStyles are applied to form', () => {
          props.flat = true
          const expectedProps = {
            borderRadius: 'unset',
            boxShadow: 'unset',
            marginTop: 'unset'
          }
          expect(comp().find('form').prop('style')).toEqual(expectedProps)
        })

        it('if false, flatStyles are not applied to form', () => {
          props.flat = false
          const expectedProps = {
            borderRadius: 'unset',
            boxShadow: 'unset',
            marginTop: 'unset'
          }
          expect(comp().find('form').prop('style')).not.toEqual(expectedProps)
        })
      })
    })
  })

  describe('functions', () => {
    describe('form', () => {
      describe('onSubmit', () => {
        it('calls preventDefault()', () => {
          const e = { preventDefault: jest.fn() }
          comp().find('form').simulate('submit', e)
          expect(e.preventDefault).toHaveBeenCalled()
        })

        it('calls props.onSubmit', () => {
          const e = { preventDefault: () => {} }
          props.onSubmit = jest.fn()
          comp().find('form').simulate('submit', e)
          expect(props.onSubmit).toHaveBeenCalled()
        })

        it('calls props.onSubmit with correct args', () => {
          const e = { preventDefault: () => {} }
          props.onSubmit = jest.fn()
          comp().find('form').simulate('submit', e)
          expect(props.onSubmit).toHaveBeenCalledWith(comp().state(), comp().instance().handleSubmit)
        })
      })
    })

    describe('textarea Field', () => {
      describe('onChange', () => {
        it('calls setState', () => {
          const e = { target: { value: 'val' } }
          const mock = jest.spyOn(Reply.prototype, 'setState')
          comp().find('field[type="textarea"]').simulate('change', e)
          expect(mock).toHaveBeenCalled()
        })

        it('calls setState w/ correct props', () => {
          const e = { target: { value: 'val' } }
          const mock = jest.spyOn(Reply.prototype, 'setState')
          comp().find('field[type="textarea"]').simulate('change', e)
          expect(mock).toHaveBeenCalledWith({ text: 'val' })
        })
      })
    })

    describe('handleSubmit()', () => {
      it('calls setState', () => {
        const spy = jest.spyOn(Reply.prototype, 'setState')
        comp().instance().handleSubmit()
        expect(spy).toHaveBeenCalled()
      })

      it('calls setState w/ correct args', () => {
        const spy = jest.spyOn(Reply.prototype, 'setState')
        comp().instance().handleSubmit()
        expect(spy).toHaveBeenCalledWith({ text: '' })
      })

      it('calls fetchUsersStats', () => {
        comp().instance().handleSubmit()
        expect(props.fetchUsersStats).toHaveBeenCalled()
      })
    })
  })

  it('matches snapshot', () => {
    expect(comp()).toMatchSnapshot()
  })
})

