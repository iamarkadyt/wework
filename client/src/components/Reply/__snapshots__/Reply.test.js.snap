// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reply matches snapshot 1`] = `
<form
  className="Reply-container"
  onSubmit={[Function]}
  style={
    Object {
      "borderRadius": "unset",
      "boxShadow": "unset",
      "marginTop": "unset",
    }
  }
>
  <p
    className="Reply-header"
  >
    Post as 
    <b>
      Name
      :
    </b>
  </p>
  <field
    onChange={[Function]}
    rows={5}
    style={
      Object {
        "fontSize": "1.3rem",
      }
    }
    type="textarea"
  />
  <field
    containerStyle={
      Object {
        "marginTop": ".5rem",
      }
    }
    label="Submit"
    type="submit"
  />
</form>
`;
